### DISCUSSION TITLE
In navigation, the route to walk to bus stop is not optimal

### DISCUSSION THREAD

**Discussion by xlucn** [osmandapp/OsmAnd#12643] [Maps Related]
In navigation -> public transportation, OsmAnd will guide the user to walk to a bus stop first. But the route OsmAnd planned is confusing:
1. Technically, the navigation does not lead the user to the bus stop, but to the nearest node on the road.
2. Even the route to walk from the start point to the nearest node is not optimal. In the screenshot below, the route is walking back and force for three times (see left image). It should only take one, which is the case if I navigate (walking) from the start point to the node of the road (see right image).

Would it be better to
1. go to point of the bus stop, and
2. from the bus stop, go to the nearest point (emphasize, this is not a node/point added in edit, but anywhere in a line) of the road.

<img width=30%, src="https://user-images.githubusercontent.com/12032219/131480109-1fc40513-4088-4b29-aee0-ee9ecd6c38c2.png"/><img width=30%, src="https://user-images.githubusercontent.com/12032219/131480100-d7995c2b-f0b4-4344-aa84-a99971b770ce.png"/><img width=30%, src="https://user-images.githubusercontent.com/12032219/131481386-0dec922f-ae2e-4a97-b706-e9917e8f8f93.png"/>

https://www.openstreetmap.org/node/8061284777#map=18/39.62393/118.17921

**Comment by vshcherb** [comment]
quite difficult indeed, does it help to have separated platform_stop=area and platform=stop itself? In sense of navigation any disconnected lines leads to such kind of weirdness, though I agree this super awkward

**Reply by xlucn** [reply]
Sorry, I don't quite get what you meant. Do I create another object with platform_stop=area and platform=stop, or two, one with each attribute? This bus route doesn't have either of those now.

Did a bit of searching, only [find that](https://wiki.openstreetmap.org/wiki/Tag:public_transport%3Dstop_area) `platform_stop=area` can be a relation including two `public_transport=platform` and two  `public_transport=stop_position`. But `public_transport=stop_position` is optional or [even useless](https://wiki.openstreetmap.org/wiki/Tag:public_transport%3Dstop_position) in most occasions, I guess the idea is navigation should find an optimal "stop position" if there is no stop_position, which is what I proposed.

Also, I couldn't find platform=* attribute for any objects.

**Reply by xlucn** [reply]
I do notice that when there is no nearby node, navigation usually gets strange:
```
node             node
-o----------------o--
          o
      platform
```
But if there is a node nearby, even without any data/attributes, it will be slightly better. But sometimes still weird.
```
node     node    node
-o--------o-------o--
          o
      platform
```

**Reply by vshcherb** [reply]
Yes I meant public_transport=platform, for routing graph there is should be no hanging points i.e. ```public_transport=stop_position``` or ```highway=stop``` should be connected to the road where the bus drives / stops and ```public_transport=platform``` should be connected to the pedestrian roads / footways where people go. I'm not sure how much it will fix the current situation but precise mapping could help to support this in the future.

As you show now ```highway=stop``` is not connected to any line at all. Though whats strange why it goes to the center of intersection cause there is no stop there...

**Reply by xlucn** [reply]
Thanks for the reply!

I read through the [OSM wiki](https://wiki.openstreetmap.org/wiki/Buses) on buses, it does not require any more elements in a bus route other than some stops, usually as nodes, and the road where the bus drives on, as lines.

IMHO, a client should work properly with those minimum data, instead of require more lines and nodes added. I admit those additional elements do help the situation, but are extremely exhausting to edit (kinda doubles the work).

In the example bus route provided in the wiki page above, all the bus stops are beside the road, with no lines connected at all, see [this node](https://www.openstreetmap.org/node/469769787#map=19/51.45190/-0.07079) as an example.

**Reply by vshcherb** [reply]
App doesn't require any additional information for Public Transport navigation, it should work correctly proposing the routes though the walking routes to public transport stops won't be correct if more information is not provided

**Reply by xlucn** [reply]
> proposing the routes though the walking routes to public transport stops

That's what I am mainly talking about in this thread (see the title). OSM instructs to [add the bus stop](https://wiki.openstreetmap.org/wiki/Buses#Bus_stops) "at the point where passengers would wait immediately prior to boarding the vehicle", I think it's reasonable for a client to automatically find a shortest, even quite possibly, *not-existing* walking route between the stop and road.

The point is, there should be no barrier or anything stopping people to walk straight from the bus stop to the road, that's by design in real life. Even if in OSM we don't add pedestrian roads between the "bus stop-position" and the platform, the app should assume there is one.

Edit: I also think that is the intention of the OSM wiki not requiring to add a path between public_transport=platform and public_transport=stop_position (or where it should be). If for almost every bus stop we should do the same thing: draw a vertical line from the stop to the road, why not make it implicit?

**Reply by vshcherb** [reply]
True, that's how navigation in OsmAnd works it find the closest suitable road and picks a closest point from it  and navigates to it. Though in your screenshots something looks weird. 

Could you please double check if there is no issue with transport stop itself? Probably public transport route relation has a wrong point. You can also examine points position in OsmAnd click on the trasport stop -> Route shield and you can navigate trhough stops
https://docs.osmand.net/en/main@latest/osmand/map/public-transport#browse-route

**Reply by xlucn** [reply]
> Could you please double check if there is no issue with transport stop itself? Probably public transport route relation has a wrong point.

There is nothing wrong in terms of member order. I can reproduce the issue in other bus routes, see below.

> find the closest suitable road and picks a closest point from it and navigates to it

I found out that's partially true. See the examples below.

## Explain my use of "point" and "node"

Before anything, it seems that the terms I use are not standard, since I don't know the terminologies that well. I will still use my choices, but to explain what I mean here:
- "node": I mean an actual point we can add in OSM editor (shortcut '1' in iD editor);
- "point": I mean *any* point, as in a mathematical or physics concept (given any coordinate on earth, there will be a "point", but not necessarily been added as a "node" in OSM), usually on a road in our topic here.

## Issue and examples for demonstration
There are other issues I found, but the following is definitely reproducible anywhere, while for the other issues I will look further later.

**Issue**: If by "a closest point" you mean the same concept of "point" I use, then I found that **OSMand does not choose the closest "point" on the road for the starting point, but the closest "node"**. And weirdly enough, the end point is usually correct.

I will use Bus 12 of London (used as [an example in OSM wiki](https://wiki.openstreetmap.org/wiki/Buses#Example_bus_routes))  here to demonstrate. It's a simple bus route without stop positions.

**Example 1**:  Goodrich Road -> Underhill Road (south to north)
Left: the navigation result; middle: edit window around "Goodrich Road" stop; right: edit window around "Underhill Road" stop.
Facts:
1. From the OSM editor window, you can see there is no "node"s near the bus stops where `public_transport=stop_position` should be if one is going to be added.
2. The navigation is
   -  first walk to the nearest "node", which is the "crossing" ⊙ south to Goodrich Road stop
   - then take the bus to Underhill Road stop
   - last get off at the nearest "point" to Underhill Road stop, where there is no "node" at all.

Summary: OSMand finds the wrong point to get on the bus, but the correct point to get off.
<img width=25%, src="https://user-images.githubusercontent.com/12032219/132276896-10c5b48e-9363-4ae7-9cb9-47d27efdad7f.png"/> <img width=34%, src="https://user-images.githubusercontent.com/12032219/132278112-3613b0c7-ab33-4947-b5b1-889740df7730.png"/> <img width=34%, src="https://user-images.githubusercontent.com/12032219/132278107-6840901e-92b8-457d-9341-0285e0d8b0b7.png"/>

**Example 2**: Etherow Street -> Goodrich Road (south to north)
Left: the navigation result; right: edit window around "Etherow Street" stop.
This is similar to example 1. Since Etherow Street is the starting point of the bus route, a nearby "node" is required to split the line in order to include the top half in the relation. Thus, that "node" is used as the "point" to get on the bus.
**Note** that the end point "Goodrich Road" is the same as in example 1. But the navigation does not lead to the crossing. This shows the difference of a location as start point and end point in a navigation.
<img width=25%, src="https://user-images.githubusercontent.com/12032219/132276889-a153598b-b0d6-43be-9ad4-1396efc7de77.png"/> <img width=34%, src="https://user-images.githubusercontent.com/12032219/132278114-67d6eb96-6e6a-4491-900e-cb363a5eb65c.png"/>

**Example 3**: Underhill Road -> Goodrich Road (north to south)
Left: the navigation result; right: edit window around "Underhill Road" stop.
Still similar to example 1. The nearest "node" is the "T" crossing, so the navigation leads to walk there.
<img width=25%, src="https://user-images.githubusercontent.com/12032219/132276898-baf706b0-39ab-4bdf-b6f5-6113e7ca1401.png"/>  <img width=34%, src="https://user-images.githubusercontent.com/12032219/132278100-a38190bc-db9b-40b1-9ddb-3cc16c811f78.png"/>

I really hope I made my point clear this time ...

As I said, this issue is not the only reason for the behavior in my first screenshot. I think: 1. not starting from the bus stop exactly and 2. complicated road condition also affect it in some way. I will look into it and find if there are any patterns.

**Reply by vshcherb** [reply]
> OSMand does not choose the closest "point" on the road for the starting point, but the closest "node"

That would be an issue though for sure for simple car / pedestrian routing it was taken into account, probably it's something wrong with Public Transport routing. I will need to double check as well
